File: vectorc2/command/commander.py
#  Copyright 2019 Sebastian Ryszard Kruk <vectorc2@kruk.me>
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
import sys
from io import StringIO
import contextlib

def rgb_to_hs(rgbstr):
  """
  Convert RGB color to (hue, saturation)
  """
  import colorsys
  import math
  r, g, b = bytes.fromhex(rgbstr[1:])
  h, s, v = colorsys.rgb_to_hsv(r, g, b)

  s2 = 0.5+(255-v)/512

  return {'hue':h, 'saturation':s2}

class OuputProxy(object):
  """
  Proxy object wrapping outputing to given stream, stringIO, and WS consumer
  """
  def __init__(self, stdout, stringio, consumer, message_type):
    self._stdout = stdout
    self._stringio = stringio
    self._type = message_type
    self._consumer = consumer

  def __getattr__(self, name):
    if name in ('_stdout', '_stringio', '_type', '_consumer', 'write'):
      object.__getattribute__(self, name)
    else:
      return getattr(self._stringio, name)

  def write(self, data):
    """
    Performs the actual write to registered outputs
    """
    self._stdout.write(data)
    self._stringio.write(data)
    if self._type is not None:
      self._consumer.send_message(data, _type=self._type)


class Commander(object):
  """
  The most important class here: enables to run code enpacked in text, and route output back to user.
  """

  def __init__(self, consumer):
    self.consumer = consumer

  @contextlib.contextmanager
  def _stdoutIO(self, stdout=None, stderr=None):
    """
    Trick to replace stdout and stderr with proxy
    """
    oldout, olderr = sys.stdout, sys.stderr
    if stdout is None:
      stdout = StringIO()
    if stderr is None:
      stderr = StringIO()
    sys.stdout = OuputProxy(sys.stdout, stdout, self.consumer, 'text')
    # sys.stderr = OuputProxy(sys.stderr, stderr, self.consumer, 'error')
    sys.stderr = OuputProxy(sys.stderr, stderr, self.consumer, None)
    yield sys.stdout, sys.stderr
    sys.stdout, sys.stderr = oldout, olderr

  def run(self, code):
    """
    Call to run code    
    """
    # TODO: add protection against imports, __* type of stuff

    with self._stdoutIO() as (sout, serr):
      try:
        exec(code)
      except:
        e = sys.exc_info()[0]
        self.consumer.send_message(e, _type="error")
      finally:
        self.consumer.send_message("finished", _type="command")

File: vectorc2/command/static/script/vector/commander.js
/**
 *  Copyright 2019 Sebastian Ryszard Kruk <vectorc2@kruk.me>
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 * @author vectorc2@kruk.me
 */
'use strict';

/**
 * Main class handling commands being sent from VectorC2 WebUI to Python ... and logging back.
 */
const Commander = (function(){

  /**
   * Reference to VectorSocket object
   */
  var _vectorSocket;

  /**
   * Temporary reference to the callback method
   */
  var __onCommandCallback = null;

  /**
   * Sends given code over to Server to execute against Vector API. 
   * Once the code is done executing _callback function is called.
   * @param {String} code 
   * @param {Function} _callback 
   */
  function _runPython(code, _callback) {
    _vectorSocket.sendMessage(code);
    __onCommandCallback = _callback;
  }

  /**
   * Run user block in JavaScript.
   * For quick and dirty test purposes
   */
  function _runJavaScript(code, _callback) {
    Blockly.JavaScript.INFINITE_LOOP_TRAP = '  checkTimeout();\n';
    var timeouts = 0;
    var checkTimeout = function() {
      if (timeouts++ > 1000000) {
        throw gettext('Maximum execution iterations exceeded.');
        _callback('timeout');
      }
    };
    Blockly.JavaScript.INFINITE_LOOP_TRAP = null;

    try {
      eval(code);
    } catch (e) {
      let msg = gettext('Program error:\n%(error)s'); 
      LogPanel.logError(interpolate(msg, {error: e}, true));
    }
    _callback('finished');
  };


  /**
   * Initializes Commander by starting WebSocket 
   */
  function __init__() {
    _vectorSocket = VectorSocket('c2', Commander);
  }

  /**
   * Called by the server
   * @param {String} message 
   */
  function _onCommand(message) {
    if (__onCommandCallback !== null) {
      __onCommandCallback(message);
      __onCommandCallback = null;
    }
  }
  // ---------------------------------------------------------------------------

  return {
      init: __init__,
      onCommand: _onCommand,
      run: {
        'js': _runJavaScript,
        'py': _runPython
      }
  }

})();

$( document ).ready(Commander.init)